### PR TITLE
feat: contracts middleware interface

### PR DESCRIPTION
**Pull request by thedevbirb**
This PR introduces the `IBoltMiddleware` interface and decouples many functionalities of the `BoltManager` contract into two separate contracts: `BoltEigenLayerMiddleware` and `BoltSymbioticMiddleware`. 
Moreover, the `BoltManager` contract is now agnostic of the available restaking protocols.

## Summary by CodeRabbit

- **New Features**
	- Introduced `BoltEigenLayerMiddleware` and `BoltSymbioticMiddleware` for enhanced integration with the Bolt Protocol.
	- Updated registration processes for operators and vaults to streamline user interactions.
	- Added functionalities for managing operator and collateral stakes, including epoch management.

- **Bug Fixes**
	- Clarified and updated integration steps in the documentation for better user understanding.

- **Refactor**
	- Simplified the `BoltManager` by removing EigenLayer and Symbiotic-related functionalities and focusing on restaking protocols. 

- **Documentation**
	- Enhanced the `README.md` to reflect changes in the integration process and middleware structure.

### CODE REVIEW

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol:67]: missing doc

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol]: missing params

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol]: name hash variable is already public, prbably don't need this one

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol]: these are symbiotic specific, not used here, can they be removed?

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol:209]: missing doc

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol]: missing a / on these lines

File: bolt-contracts/src/contracts/BoltManager.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltManager.sol:1]: should we keep track of whitelisted assets by protocol or in a single mapping?
e.g. `symbioticWhitelistedCollaterals` or just `whitelistedCollaterals`? 

In theory once a collateral token is approved it should be used anywhere, if possible. 
Just an observation

File: bolt-contracts/src/interfaces/IBoltMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/interfaces/IBoltMiddleware.sol:34]: why is this part of the middleware? IMO handling collateral whitelist should be a responsibility of the manager governance, wdyt?

**Review by merklefruit — COMMENTED** [review]
Really like the direction

File: bolt-contracts/src/contracts/BoltManager.sol
Comment (thedevbirb) [reply bolt-contracts/src/contracts/BoltManager.sol:1]: I'd love a more abstracted mapping but I think it makes more sense to keep them closer to the restaking protocol since they might have different definition of "collateral". In the context of EigenLayer we are fine with whitelisted ERC20 addresses but in Symbiotic you need to specify the contract instances of "Collateral".

Having a single mapping might be ambiguous because you might think you can use any of such collaterals in any restaking protocol but that's not the case. 

What do you think? Not trying to force a point but I think we should genuinely analyse this further

File: bolt-contracts/src/interfaces/IBoltMiddleware.sol
Comment (thedevbirb) [reply bolt-contracts/src/interfaces/IBoltMiddleware.sol:34]: Consider also [this discussion](https://github.com/chainbound/bolt/pull/235#discussion_r1770990832) I think it makes sense that every middleware must implement its own logic to manage the supported collaterals.

File: bolt-contracts/src/contracts/BoltManager.sol
Comment (merklefruit) [reply bolt-contracts/src/contracts/BoltManager.sol:1]: > Having a single mapping might be ambiguous because you might think you can use any of such collaterals in any restaking protocol but that's not the case.

totally agree with this, however on the flip side having separate whitelists means having governance specific to the middleware instead of keeping all bolt governance stuff on the `BoltManager`. It doesn't make a huge difference except complicating things a little bit...

**Comment by thedevbirb** [comment]
@coderabbitai review

**Comment by thedevbirb** [comment]
@coderabbitai review

File: bolt-contracts/src/interfaces/IBoltMiddleware.sol
Comment (thedevbirb) [reply bolt-contracts/src/interfaces/IBoltMiddleware.sol:35]: wdyt @merklefruit? I think it could be helpful

File: bolt-contracts/test/BoltManager.EigenLayer.t.sol
Comment (thedevbirb) [reply bolt-contracts/test/BoltManager.EigenLayer.t.sol]: I think the `ProposerStatus` type should be defined in `BoltValidators`. Wydt @merklefruit?

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (thedevbirb) [reply bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol:382]: Yeah this is bad and implies a contract upgrade once slashing is live on EL. @merklefruit do we want to separate storage contract from implementation contracts and change some implementations using governance? Can't think of a better way.
Or perhaps a storage switch like the one used for BLS

File: bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol]: now we can just call these `operators` and `vaults`.
the same with immutables below: we don't need SYMBIOTIC prefix anymore

File: bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol]: we should read this address from `BoltManager` e.g. `IBoltManager.validators` to make it upgradable in a single place. Then we just add a reference to BoltManager in all adapters

File: bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol:418]: I'll track this in an issue right now!

Comment (merklefruit) [reply bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol:418]: https://github.com/chainbound/bolt/issues/238

File: bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol
Comment (thedevbirb) [reply bolt-contracts/src/contracts/BoltSymbioticMiddleware.sol]: We will handle it in a future PR, resolving this

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol]: if `validators` is now defined in `BoltManager`, can we access it by storing just the `BoltManager` address instead of the `BoltValidators` one here?

File: bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol
Comment (merklefruit) [inline bolt-contracts/src/contracts/BoltEigenLayerMiddleware.sol]: It would be great to use the interface instead of the contract here for consistency

**Review by merklefruit — APPROVED** [review]
LGTM

**Pull request merged**
